sdk/python/lib/pulumi/provider/experimental/property_value.py

```diff
@@ -0,0 +1,563 @@
+        is_secret: bool = False,
+        dependencies: Optional[Iterable[str]] = None,
+    ) -> None:
+        """
+        :param value: The value of the property.
+        :param is_computed: Whether the value is computed.
+        """
+        # Wrap Sequence and Mapping types in immutable types to ensure they are hashable.
+        if isinstance(value, Sequence) and not isinstance(value, str):
+            value = tuple(value)
+        elif isinstance(value, Mapping):
+            value = types.MappingProxyType(value)
+        elif isinstance(value, pulumi.Asset) or isinstance(value, pulumi.Archive):
+            # Ensure the Asset/Archive is immutable by deep copying
+            value = copy.deepcopy(value)
+        object.__setattr__(self, "value", value)
+        object.__setattr__(self, "is_secret", is_secret)
+        object.__setattr__(
+            self,
+            "dependencies",
+            frozenset(dependencies) if dependencies else frozenset(),
+        )
+
+    @staticmethod
+    def computed() -> "PropertyValue":
+        """
+        Creates a computed property value.
+        """
+        return PropertyValue(Computed())
+
+    @staticmethod
+    def null() -> "PropertyValue":
+        """
+        Creates a null property value.
+        """
+        return PropertyValue(None)
+
+    def with_secret(self, is_secret: bool) -> "PropertyValue":
+        """
+        Returns a copy of the PropertyValue with the specified secret status.
+        """
+        return PropertyValue(
+            self.value,
+            is_secret=is_secret,
+            dependencies=self.dependencies,
+        )
+
+    def with_dependencies(self, dependencies: Iterable[str]) -> "PropertyValue":
+        """
+        Returns a copy of the PropertyValue with the specified dependencies.
+        """
+        return PropertyValue(
+            self.value,
+            is_secret=self.is_secret,
+            dependencies=dependencies,
+        )
+
+    def with_value(self, value: PythonValue) -> "PropertyValue":
+        """
+        Returns a copy of the PropertyValue with the specified value.
+        """
+        return PropertyValue(
+            value,
+            is_secret=self.is_secret,
+            dependencies=self.dependencies,
+        )
+
+    @property
+    def type(self) -> PropertyValueType:
+        """
+        Determines the type of the property value.
+        """
+        if self.value is None:
+            return PropertyValueType.NULL
+        if isinstance(self.value, bool):
+            return PropertyValueType.BOOL
+        if isinstance(self.value, (int, float)):
+            return PropertyValueType.NUMBER
+        if isinstance(self.value, str):
+            return PropertyValueType.STRING
+        if isinstance(self.value, pulumi.Asset):
+            return PropertyValueType.ASSET
+        if isinstance(self.value, pulumi.Archive):
+            return PropertyValueType.ARCHIVE
+        if isinstance(self.value, list):
```

Comment:
I think this is wrong since we convert sequences to tuples. Should we check for `Sequence` or even `tuple` here?
```suggestion
        if isinstance(self.value, Sequence):
```

Reply:
Ack yeh this was wrong, added a test and fixed